# Data Wrangling

En la fase anterior del proyecto extrajimos informacion financiera utilizando una API. Como la API nos limitaba la cantidad de llamadas diarias, cada dia habia que ajustar el codigo para que nos trajera informacion de nuevas empresas, y ademas se tuvo que ajustar de que índices bursátil provienen cada empresa. Esto me permitio recopilar 228 empresas que provienen de los indices del NASDAQ, Dow Jones y algunas empresas del S&P500.
No obstante, esta informacion debe ser purificada por diversas razones:
- Varias empresas se encuentran en mas de un indice, por lo que habra muchos valores repetidos.
- El hecho de tener que modificar el codigo diariamente puede generar datos con errores humanos
- La API puede no disponer de todos los datos que le solisitamos para todas las empresas

Es por esto que nos encargaremos de depurar los datos que esten repetidos o sean invalidos.


In [20]:
import pandas as pd

In [21]:
API_dataset = pd.read_csv('https://media.githubusercontent.com/media/Gonzalo-Ariel-Alonso/Proyecto_Data_Science/master/combined_API_dataset.csv',index_col = 0)
API_dataset

Debt ratio %  Current Liability to Assets   
ATVI      0.131822                     0.000000  \
ADBE      0.147748                     0.021609   
ADP       0.055454                     0.002163   
ABNB      0.116695                     0.003679   
ALGN      0.021483                     0.004468   
...            ...                          ...   
DUK       0.430676                     0.045473   
SO        0.449331                     0.052568   
ADSK      0.295953                     0.009006   
ICE       0.121122                     0.000021   
USB       0.144964                     0.046259   

      Borrowing dependency (Debt To Equity)   
ATVI                               0.179518  \
ADBE                               0.277194   
ADP                                0.898993   
ABNB                               0.441504   
ALGN                               0.036430   
...                                     ...   
DUK                                1.563520   
SO                                 1.983425   
ADSK                               2.992196   
ICE                                0.792334   
USB                                1.866803   

      Current Liability to Current Assets  Net Income to Total Assets   
ATVI                             0.000000                    0.055253  \
ADBE                             0.065251                    0.175078   
ADP                              0.002489                    0.046757   
ABNB                             0.003970                    0.118032   
ALGN                             0.010961                    0.060790   
...                                   ...                         ...   
DUK                              0.613069                    0.014518   
SO                               0.680780                    0.025413   
ADSK                             0.025441                    0.087201   
ICE                              0.000025                    0.007708   
USB                              0.228633                    0.008632   

      Net worth/Assets  Persistent EPS in the Last Four Seasons (EPS)   
ATVI          0.702735                                           1.92  \
ADBE          0.517246                                          10.13   
ADP           0.051140                                           7.04   
ABNB          0.346677                                           3.07   
ALGN          0.605479                                           4.62   
...                ...                                            ...   
DUK           0.277352                                           5.08   
SO            0.225426                                           3.28   
ADSK          0.121318                                           3.81   
ICE           0.116838                                           2.59   
USB           0.075231                                           3.86   

      Working Capital/total assets  
ATVI                      0.427345  
ADBE                      0.046199  
ADP                      -0.000284  
ABNB                      0.415077  
ALGN                      0.055025  
...                            ...  
DUK                      -0.020762  
SO                       -0.032159  
ADSK                     -0.115808  
ICE                       0.042863  
USB                       0.112394  

[453 rows x 8 columns]

## Datos repetidos

In [22]:
sum(API_dataset.duplicated())

284

Podemos observar que tenemos 284 filas repetidas, se procede a eliminarlas.

In [23]:
API_dataset.drop_duplicates(inplace = True)

## Datos Nulos
Contamos que cantidad de datos nulos tenemos

In [24]:
API_dataset.isnull().sum().sum()

4

Existen 4 datos nulos, busquemos en que filas se encuentran

In [25]:
API_dataset[API_dataset.isnull().any(axis = 1)]

Debt ratio %  Current Liability to Assets   
BA       0.40625                     0.037856  \

    Borrowing dependency (Debt To Equity)   
BA                              -3.571769  \

    Current Liability to Current Assets  Net Income to Total Assets   
BA                             0.047387                         NaN  \

    Net worth/Assets  Persistent EPS in the Last Four Seasons (EPS)   
BA               NaN                                            NaN  \

    Working Capital/total assets  
BA                           NaN

Podemos observar que los 4 valores nulos se encuentran en la misma fila, por lo tanto eliminar una sola fila no supone un impacto mayor en nuestros dataset

In [26]:
API_dataset.dropna(inplace = True)

## Datos invalidos
Nuestro dataset se compone de mayormente de indices economicos, por lo tanto aquellos datos que tengan valor 0 los trataremos como datos invalidos

In [27]:
API_dataset[(API_dataset == 0).any(axis=1)]

Debt ratio %  Current Liability to Assets   
ATVI      0.131822                     0.000000  \
AMZN      0.304388                     0.000000   
ANSS      0.116288                     0.000000   
BIIB      0.268704                     0.000000   
FTNT      0.145032                     0.000000   
ISRG      0.000000                     0.000000   
LRCX      0.260109                     0.000000   
MNST      0.000000                     0.000000   
NFLX      0.291716                     0.000000   
NXPI      0.470546                     0.000000   
PYPL      0.135789                     0.000000   
SGEN      0.025718                     0.000000   
IBM       0.461519                     0.044277   
LMT       0.285599                     0.000000   
PNC       0.108267                     0.000000   

      Borrowing dependency (Debt To Equity)   
ATVI                               0.179518  \
AMZN                               0.914739   
ANSS                               0.157597   
BIIB                               0.479415   
FTNT                              86.921053   
ISRG                               0.000000   
LRCX                               0.595355   
MNST                               0.000000   
NFLX                               0.661398   
NXPI                               1.404831   
PYPL                               0.527797   
SGEN                               0.033463   
IBM                                2.854842   
LMT                                1.617251   
PNC                                1.240152   

      Current Liability to Current Assets  Net Income to Total Assets   
ATVI                             0.000000                    0.055253  \
AMZN                             0.000000                   -0.005883   
ANSS                             0.000000                    0.078307   
BIIB                             0.000000                    0.120615   
FTNT                             0.000000                    0.137540   
ISRG                             0.000000                    0.103623   
LRCX                             0.000000                    0.239318   
MNST                             0.000000                    0.143689   
NFLX                             0.000000                    0.092436   
NXPI                             0.000000                    0.121923   
PYPL                             0.000000                    0.030730   
SGEN                             0.000000                   -0.166091   
IBM                              0.193489                    0.000000   
LMT                              0.000000                    0.108396   
PNC                              0.000000                    0.010970   

      Net worth/Assets  Persistent EPS in the Last Four Seasons (EPS)   
ATVI          0.702735                                           1.92  \
AMZN          0.315649                                          -0.27   
ANSS          0.727555                                           6.02   
BIIB          0.545261                                          20.96   
FTNT         -0.045215                                           1.05   
ISRG          0.856528                                           3.71   
LRCX          0.436898                                          32.92   
MNST          0.847094                                           1.13   
NFLX          0.427565                                          10.10   
NXPI          0.320580                                          10.64   
PYPL          0.257556                                           2.10   
SGEN          0.763041                                          -3.30   
IBM           0.172457                                           1.82   
LMT           0.175227                                          20.74   
PNC           0.082141                                          13.43   

      Working Capital/total assets  
ATVI                      0.427345  
AMZN                    

In [28]:
API_dataset[(API_dataset == 0).any(axis=1)].shape

(15, 8)

Tenemos 15 filas en las cuales al menos existe un valor invalido, posiblemente porque en los balances de estas empresas no se precentes estos datos economicos en especifico, asi que se procede a eliminarlas.

In [29]:
API_dataset.drop(API_dataset[API_dataset.eq(0).any(axis=1)].index , inplace = True)

## Valores atipicos

Los outliers pueden ser un problema a la hora de hacer un modelo preciso, es por eso que eliminaremos aquellos valores que esten fuera los percentiles de 10% y 90%

In [30]:
# Calcular el rango intercuartílico (IQR)
P10 = API_dataset.quantile(0.10)
P90 = API_dataset.quantile(0.90)
P = P90 - P10

# Detectar filas con valores atípicos
outliers = ((API_dataset < (P10 - 1.5 * P)) | (API_dataset > (P90 + 1.5 * P))).any(axis=1)

# Filtrar y mostrar las filas con valores atípicos
filas_atipicas = API_dataset[outliers]
filas_atipicas

Debt ratio %  Current Liability to Assets   
AMGN      0.694263                     0.024431  \
BKNG      0.502460                     0.019715   
CHTR      0.676275                     0.010448   
IDXX      0.428304                     0.238092   
ILMN      0.252561                     0.101861   
MAR       0.469815                     0.027564   
ORLY      0.551743                     0.029040   
PANW      0.278508                     0.300059   
VRSK      0.735203                     0.204336   
MCD       0.959719                     0.013108   
HD        0.647082                     0.028465   
PM        0.759281                     0.133720   
ORCL      0.673302                     0.030219   
SPG       0.782801                     0.015084   
AMT       0.702045                     0.078922   
MO        0.689377                     0.042106   
PLD       0.284152                     0.017503   

      Borrowing dependency (Debt To Equity)   
AMGN                              11.517390  \
BKNG                              11.792365   
CHTR                              10.402633   
IDXX                               1.430087   
ILMN                               0.445690   
MAR                               83.435714   
ORLY                              -4.404669   
PANW                               3.202369   
VRSK                              45.036550   
MCD                               -8.642371   
HD                               136.541436   
PM                                -5.280255   
ORCL                              58.149743   
SPG                                8.678351   
AMT                                8.676248   
MO                                -6.549794   
PLD                                0.476072   

      Current Liability to Current Assets  Net Income to Total Assets   
AMGN                             0.071712                    0.100613  \
BKNG                             0.031650                    0.120579   
CHTR                             0.375902                    0.040471   
IDXX                             0.593732                    0.247232   
ILMN                             0.350463                   -0.359452   
MAR                              0.206459                    0.095023   
ORLY                             0.072643                    0.172050   
PANW                             0.573166                   -0.021790   
VRSK                             1.537564                    0.137090   
MCD                              0.121880                    0.122481   
HD                               0.067014                    0.223756   
PM                               0.420409                    0.154456   
ORCL                             0.193344                    0.063274   
SPG                              0.344564                    0.074289   
AMT                              1.464015                    0.025251   
MO                               0.215512                    0.155978   
PLD                              1.899944                    0.040449   

      Net worth/Assets  Persistent EPS in the Last Four Seasons (EPS)   
AMGN          0.056218                                          12.18  \
BKNG          0.109696                                          76.70   
CHTR          0.063097                                          31.30   
IDXX          0.221620                                           8.12   
ILMN          0.538606                                         -28.05   
MAR           0.022889                                           7.27   
ORLY         -0.084000                                          33.75   
PANW          0.017138                                          -0.90   
VRSK          0.251296                                           6.04   
MCD          -0.119031                                           8.45   
HD            0.020433                                          16.74   
PM           -0.145215                           

In [31]:
# Eliminar outliers
API_dataset.drop(filas_atipicas.index , inplace = True)

## Ordenar dataset
por ultimo ordenaremos nuestro dataset por orden alfabetico para que quede mejor presentado.

In [33]:
API_dataset.sort_index(inplace = True)
API_dataset

Debt ratio %  Current Liability to Assets   
AAPL      0.330007                     0.059843  \
ABBV      0.461507                     0.029797   
ABNB      0.116695                     0.003679   
ABT       0.225328                     0.030240   
ACN       0.061766                     0.015166   
...            ...                          ...   
WMT       0.260625                     0.027122   
XEL       0.427303                     0.035644   
XOM       0.112207                     0.001718   
ZM        0.010676                     0.002804   
ZS        0.379534                     0.009214   

      Borrowing dependency (Debt To Equity)   
AAPL                               1.763490  \
ABBV                               4.677791   
ABNB                               0.441504   
ABT                                0.225993   
ACN                                0.122223   
...                                     ...   
WMT                                0.882080   
XEL                                1.548460   
XOM                                0.208602   
ZM                                 0.013968   
ZS                                 2.020340   

      Current Liability to Current Assets  Net Income to Total Assets   
AAPL                             0.155903                    0.282924  \
ABBV                             0.145311                    0.085336   
ABNB                             0.003970                    0.118032   
ABT                              0.089240                    0.093138   
ACN                              0.033167                    0.147873   
...                                   ...                         ...   
WMT                              0.086979                    0.046382   
XEL                              0.423989                    0.028372   
XOM                              0.006494                    0.156007   
ZM                               0.003585                    0.012760   
ZS                               0.011566                   -0.137778   

      Net worth/Assets  Persistent EPS in the Last Four Seasons (EPS)   
AAPL          0.143646                                           6.15  \
ABBV          0.124304                                           6.65   
ABNB          0.346677                                           3.07   
ABT           0.492840                                           3.91   
ACN           0.467731                                          10.87   
...                ...                                            ...   
WMT           0.315017                                           4.29   
XEL           0.272521                                           3.17   
XOM           0.528492                                          13.26   
ZM            0.763602                                           0.35   
ZS            0.202389                                          -2.77   

      Working Capital/total assets  
AAPL                     -0.020303  
ABBV                     -0.007939  
ABNB                      0.415077  
ABT                       0.130780  
ACN                       0.133032  
...                            ...  
WMT                      -0.069803  
XEL                      -0.023011  
XOM                       0.082798  
ZM                        0.581170  
ZS                        0.437082  

[136 rows x 8 columns]